### PR TITLE
[WebThing] Initial contribution (replaces former PR#9185)

### PR DESCRIPTION
**Pull request by grro**
Signed-off-by: Gregor Roth <gregor.roth@web.de>

This merge request includes a WebThing binding for openHAB. **It replaces the earlier PR #9185** which has been closed due to underlying git problems. 

The WebThing binding supports an interface to remote devices implementing the Web Thing API. The Web Thing API is specified by the W3C Consortium (draft). The specification describes how a device or thing can be discovered and linked based on web technologies.
Essentially, a Web Thing API is a RESTful interface including WebSockets support to provide real-time notifications of events as soon as they happen.

Please refer to the readme.md for more detailed information

### CODE REVIEW

**Comment by andrewfg** [comment]
@grro this is very interesting. It occurs to me that this binding is really only half the story; it is something that imports WoT things into OH. The other half of the story would be something that exports OH Things as WoT things. Probably such an WoT exporter would need to be implemented as part of the OH core rather than as an OH add-on (since for example it would need to access the Thing registry). It might for example be an extension of the REST server. So I wonder if your WoT import binding needs any extensions or adaptions to its architecture to prepare the way for a future WoT export add-on? Do you have any thoughts about this?

**Comment by grro** [comment]
> @grro this is very interesting. It occurs to me that this binding is really only half the story; it is something that imports WoT things into OH. The other half of the story would be something that exports OH Things as WoT things. Probably such an WoT exporter would need to be implemented as part of the OH core rather than as an OH add-on (since for example it would need to access the Thing registry). It might for example be an extension of the REST server. So I wonder if your WoT import binding needs any extensions or adaptions to its architecture to prepare the way for a future WoT export add-on? Do you have any thoughts about this?


Hi, 

this PR supports consuming WebThings (as a client) only. My main motivation behind this was to reach a better integration of my pi-based hardware solutions that provides the WebThing API. Providing a WebThing binding could help makers to make their solutions more interoperable by implementing a lightweight standard, for instance, such as in my case.

Providing openHAB artifacts via the WebThings API would also be nice. However, my impression is, that the need for consuming support is higher. Exposing WebThings may be a second step behind this PR. I would prefer to implement features step by step in an iterative, incremental way. 
 
Essentially, this binding implements (1) a client-side interface similar to https://www.w3.org/TR/wot-scripting-api/#the-consumedthing-interface, (2) the WebThing discovery, (3) the (semantic) mapping of WebThing properties <-> openHAB channels as well as (4) the integration into the openHAB infrastructure. At least (3) the (semantic) mapping of WebThing properties <-> openHAB channels should be reusable by providing the server-side support. I have tried to keep the implementation design as clean (-> separation of concerns, encapsulation dependencies, etc) as possible which may help

**Comment by andrewfg** [comment]
^
Very interesting. Thank you.

File: bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/WebThingHandler.java:182]: You should specify the concrete exception you want to catch, since catching `Exception` catches also all runtime exceptions.
```suggestion
            } catch (IllegalArgumentException illegalURIException) {
```

File: bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/ConsumedThingImpl.java
Comment (fwolter) [inline bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/ConsumedThingImpl.java:137]: When is this exception raised? If I see correctly it is not cought anywhere. That means your binding will crash if it is raised.

File: bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java
Comment (fwolter) [inline bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java:124]: Syntactical sugar.
```suggestion
        scheduler.submit(() -> scan(isBackground));
```

File: bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java
Comment (fwolter) [inline bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/discovery/WebthingDiscoveryService.java:274]:
```suggestion
            Map<String, Object> properties = new HashMap<>(2);
```

**Review by fwolter — CHANGES_REQUESTED** [review]
There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

File: bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/ConsumedThingImpl.java
Comment (grro) [reply bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/ConsumedThingImpl.java]: The exception will be thrown if the webthing does not support a web socket uri (which should). I have changed the exception type to IOException which will be handled inside the WebThingHandler class. Thanks for the hint

**Comment by grro** [comment]
> There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

The warning are NullAnnotationsCheck warnings.

I understood that this check does not make sense for classes such as WebThingConfiguration. Additionally, the NullAnnotationsCheck for test classes seems to be a slight exaggeration in my opinion. What do you think? 
I have also tried to suppress the checkstyle rules for these cases without success

**Comment by fwolter** [comment]
The null annotations can be omitted for DTOs. The configuration should be annotated as the values can be null if the user forgets to specify a parameter in textual configuration.

The annotations are also missing for PropertyAccessException, WebSocketConnectionImpl and UnknownPropertyException.

**Comment by grro** [comment]
> The null annotations can be omitted for DTOs. The configuration should be annotated as the values can be null if the user forgets to specify a parameter in textual configuration.
> 
> The annotations are also missing for PropertyAccessException, WebSocketConnectionImpl and UnknownPropertyException.

checkstyle warnings are fixed now

File: bundles/pom.xml
Comment (fwolter) [inline bundles/pom.xml:307]: There are trailling spaces.

File: bundles/org.openhab.binding.webthing/src/main/feature/feature.xml
Comment (fwolter) [inline bundles/org.openhab.binding.webthing/src/main/feature/feature.xml:15]: The features.xml don't need a license header.

File: bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/ChannelHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/ChannelHandler.java:2]: You can run mvn license:format to update the year.

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/pom.xml
Comment (grro) [reply bundles/pom.xml]: fixed

File: bundles/org.openhab.binding.webthing/src/main/feature/feature.xml
Comment (grro) [reply bundles/org.openhab.binding.webthing/src/main/feature/feature.xml]: removed

File: bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/ChannelHandler.java
Comment (grro) [reply bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/ChannelHandler.java]: done

File: bundles/org.openhab.binding.webthing/src/main/feature/feature.xml
Comment (grro) [reply bundles/org.openhab.binding.webthing/src/main/feature/feature.xml]: header removed

**Review by fwolter — APPROVED** [review]
LGTM

Since this is a new binding, another maintainer needs to review your code before it can be merged.

**Comment by grro** [comment]
> LGTM
> 
> Since this is a new binding, another maintainer needs to review your code before it can be merged.

Thank you

**Comment by grro** [comment]
> > LGTM
> > Since this is a new binding, another maintainer needs to review your code before it can be merged.
> 
> Thank you

Do I have to do anything regarding the second reviewer?

**Comment by fwolter** [comment]
You don't need to as the "cre" tag signals a pending second review, but you can additionally request a review from `openhab/add-ons-maintainers`.

**Comment by grro** [comment]
> You don't need to as the "cre" tag signals a pending second review, but you can additionally request a review from `openhab/add-ons-maintainers`.

I have provided a bugfix commit. It turned out that there are webthing descriptions that include several "alternate" links. 

Furthermore, I noticed that may local build reports "Header line doesn't match pattern ^ \* Copyright \(c\) 2010-2020 Contributors to the openHAB project$" errors. These errors have not appeared in my local build of the previous commit (9 days ago).

**Comment by fwolter** [comment]
Your changes look good. To fix the build, you need to rebase your code to get the latest checkstyle rules.

Here are the commands for rebasing your branch:

If not already done, add the upstream openHAB addon repo as a remote to your local repo and fetch it:
```
git remote add upstream https://github.com/openhab/openhab-addons.git
git fetch upstream
```
Then, you can rebase your PR's branch onto `main`:
```
git rebase upstream/main
```
Finally force-push the rebased branch to this PR's branch:
```
git push origin [your branch name of this PR] --force-with-lease
```

**Comment by grro** [comment]
> Your changes look good. To fix the build, you need to rebase your code to get the latest checkstyle rules.
> 
> Here are the commands for rebasing your branch:
> 
> If not already done, add the upstream openHAB addon repo as a remote to your local repo and fetch it:
> 
> ```
> git remote add upstream https://github.com/openhab/openhab-addons.git
> git fetch upstream
> ```
> 
> Then, you can rebase your PR's branch onto `main`:
> 
> ```
> git rebase upstream/main
> ```
> 
> Finally force-push the rebased branch to this PR's branch:
> 
> ```
> git push origin [your branch name of this PR] --force-with-lease
> ```

unfortunately, this results in DCO problems

**Comment by fwolter** [comment]
I'm wondering what causes these rebasing fails. Can you paste the output of `git remote --verbose`?

**Comment by grro** [comment]
> I'm wondering what causes these rebasing fails. Can you paste the output of `git remote --verbose`?

git remote --verbose
origin  https://github.com/gregorroth/openhab-addons.git (fetch)
origin  https://github.com/gregorroth/openhab-addons.git (push)
upstream        https://github.com/openhab/openhab-addons.git (fetch)
upstream        https://github.com/openhab/openhab-addons.git (push)

**Comment by fwolter** [comment]
That looks good. And `git status`?

**Comment by grro** [comment]
> That looks good. And `git status`?

git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean

**Comment by fwolter** [comment]
The commit message of ee4cfe5 20min ago states that a merge occurred. I'm wondering how this happened. Are you sure you invoked `git rebase ...` and didn't do a merge via some Git UI?

**Comment by grro** [comment]
> The commit message of [ee4cfe5](https://github.com/openhab/openhab-addons/commit/ee4cfe5539f00c01aa517c8bdcb006b10c2ae26d) 20min ago states that a merge occurred. I'm wondering how this happened. Are you sure you invoked `git rebase ...` and didn't do a merge via some Git UI?

I tried to merge the change on the DescriptionTest copyright header which fails (partly?). Any idea how I can fix this mess

**Comment by fwolter** [comment]
You can make the last commit (the merge) unhappen with `git reset HEAD~1 --hard`. And then force push it with`git push --force`. Make a backup of your local repo beforehand in case anything fails.

**Comment by grro** [comment]
> You can make the last commit (the merge) unhappen with `git reset HEAD~1 --hard`. And then force push it with`git push --force`. Make a backup of your local repo beforehand in case anything fails.

worked :-) Thanks!

**Comment by grro** [comment]
will there be a second review? is there no interest in contributing this?

**Comment by fwolter** [comment]
> will there be a second review? is there no interest in contributing this?

You could request a review from @openhab/add-ons-maintainers.

**Comment by grro** [comment]
> > will there be a second review? is there no interest in contributing this?
> 
> You could request a review from @openhab/add-ons-maintainers.

Unfortunately, the link returns 404.

**Comment by bobadair** [comment]
> Unfortunately, the link returns 404.

That's weird. If you click on "reviewers" in the upper right corner and start typing "openhab" you should see openah/add-ons-maintainers come up. You shouldn't really need to do that, though. It just takes a while for another review sometimes. It looks like there are a lot of new bindings in the queue right now, which is good news for openHAB but a lot of work for the maintainers!

**Comment by moonthug** [comment]
Will this PR be reviewed anytime soon? Interested in this functionality in OpenHAB

**Comment by fwolter** [comment]
Due to lack of review capacity, the rules for new bindings were weakened. New bindings need only the approval of one reviewer instead of two from now on.

**Pull request merged**
